Sources/Basic/TerminalController.swift

```diff
@@ -0,0 +1,252 @@
+    private let fd: Int32
+    private let file: UnsafeMutablePointer<FILE>
+
+    init(file: UnsafeMutablePointer<FILE>) {
+        self.file = file
+        self.fd = fileno(file)
+    }
+
+    public func write(_ string: String) {
+        _ = fputs(string, file)
+    }
+}
+
+public var stdout = OutputFileStream(file: libc.stdout)
+public var stderr = OutputFileStream(file: libc.stderr)
+
+/// Terminal color choices.
+public enum TermColor {
+    case noColor
+
+    case red
+    case green
+    case yellow
+
+    case white
+    case black
+    case grey
+
+    /// Returns the color code which can be prefixed on a string to display it in that color.
+    /// Note: To avoid unexpected behavior the color codes are private and not exposed to clients
+    /// other than TerminalController.
+    private var string: String {
+        switch self {
+        case .noColor: return ""
+        case .red: return "\u{001B}[31m"
+        case .green: return "\u{001B}[32m"
+        case .yellow: return "\u{001B}[33m"
+        case .white: return "\u{001B}[37m"
+        case .black: return "\u{001B}[30m"
+        case .grey: return "\u{001B}[30;1m"
+        }
+    }
+}
+
+/// Number of the columns in the current terminal.
+// FIXME: What if user resizes terminal during execution?
+public var terminalWidth: Int = {
+    // Try to get from enviornment.
+    if let columns = POSIX.getenv("COLUMNS"), let width = Int(columns) {
+        return width
+    }
+
+    // Try determining using ioctl.
+    var ws = winsize()
+    if ioctl(1, UInt(TIOCGWINSZ), &ws) == 0 {
+        return Int(ws.ws_col)
+    }
+
+    // If we can't determine, just assume a default.
+    return 80
+}()
+
+/// A class to have better control on output streams especially standard output and standard error streams.
+/// Allows operations like cursor movement and colored text output if allowed by the tty.
+public final class TerminalController<Target: OutputStream> {
+
+    /// A "dumb" terminal indicates that it has no capabailities for eg: update current line, colors
+    /// The output stream is probably a file or similar.
+    public let dumb: Bool
+
+    /// Pointer to output stream to operate on.
+    private var stream: UnsafeMutablePointer<Target>
+
+    public init(stream: UnsafeMutablePointer<Target>) {
+        if case let stdStream as OutputFileStreamProtocol = stream.pointee {
+            dumb = !(isatty(stdStream.fd) == 0)
+        } else {
+            dumb = true
+        }
+
+        self.stream = stream
+    }
+
+    /// Code to clear the line on a tty.
+    private let clearLineString = "\u{001B}[2K"
+
+    /// Code to end any currently active wrapping.
+    private let resetString = "\u{001B}[0m"
+
+    /// Clears the current line and moves the cursor to beginning of the line..
+    public func clearline() {
+        guard !dumb else { return }
+        print(clearLineString + "\r", terminator: "", to: &stream.pointee)
```

Comment:
I think this class should have a single function it uses to do the bulk of the writing (and probably should take a byte string, not a String). That can embedded the `terminator: "", to: &stream.pointee` part and then the rest of the code gets cleaner.